### DISCUSSION TITLE
Inverting a multi-polygon

### DISCUSSION THREAD

**Discussion by dm356** [google/s2geometry#370] [Q&A] [answered]
I'm noticing that if I Invert a polygon containing multiple disjoint loops, the algorithm only inverts one of the loops.  Is this a bug, or is there a better way to turn three separate loops into a single polygon of three separate holes?

**Comment by smcallis** [comment, marked as answer]
Part of being valid is having polygon loops be correctly oriented relative
to one another, so holes can only exist inside a shell.  Taking three
shells and inverting one of them will flip it to contain the other two
holes and you'll end up with three total.   There's no such creature as a
valid polygon with just three holes.

**Reply by dm356** [reply]
What if I wanted to represent all of the area except what is in those three loops?  Can't you invert a single shell and get a 'hole in the world' type of concept.  What if you then subtract two more polygons from it?  Won't you get three holes with an implicit shell being full coverage of the world? Is there maybe a way of making that implicit full shell explicit?

**Reply by smcallis** [reply]
Inverting a single shell is just another shell that covers the complement of the sphere.  Adding the three shells and inverting should be doing what you want.  E.g. three shells: 

![image](https://github.com/google/s2geometry/assets/804394/c64231a0-0311-4f81-8fe0-591c7a8841ab)

and after calling Invert():

![image](https://github.com/google/s2geometry/assets/804394/480dc96e-2796-425c-9abd-6b5807a65b40)

**Reply by dm356** [reply]
Hmm. That's not what I'm seeing on my end. Maybe I'm not setting it up right. Do you have a code snippet for how you generated the above polygons?

**Comment by smcallis** [comment]
I just built a polygon and inverted it:

```
  auto polygon = s2textformat::MakePolygonOrDie(
    "16.75:-11.47,8.00:-21.84,-7.24:-12.35,6.26:1.18,16.75:-11.47;"
    "12.15:16.30,-6.54:9.44,-5.14:23.33,12.15:16.30;"
    "11.46:35.63,-9.50:36.86,-7.93:54.09,12.50:52.16,11.46:35.63");

  polygon->Invert();
```

**Reply by dm356** [reply]
Hmm. Maybe I'm just overcomplicating it on my end. I'll try it again. Thanks.

**Reply by dm356** [reply]
Ah, I was expecting is_hole to return true for the compliment. IsNormalized will better indicate the 'holes in the world' situation as I expect. Thanks for the insight.

**Reply by smcallis** [reply]
No problem!  Some of this stuff's pretty subtle.
